### PR TITLE
Add support to ExportedProgram as input to torch.onnx.dynamo_export

### LINKED ISSUE #109889: Support torch.export.export through torch.onnx.dynamo_export
### 🐛 Describe the bug

Currently, `torch.onnx.dynamo_export` leverages `torch._dynamo.export` to obtain a `GraphModule` graph. By leveraging `torch.export`, we may have a more stable and complete `GraphModule` graph, including:

* Functionalization.
* Lowering from torch ops to aten ops (aten ops are a lot more normalized than torch ops)
* Decomposition of Aten ops to a smaller opset (core aten ops)

@SherlockNoMad  FYI

### Versions

pytorch main
```[tasklist]
### Tasks
- [ ] https://github.com/microsoft/onnxscript/issues/1077
- [ ] https://github.com/pytorch/pytorch/issues/110100
```

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111498
* __->__ #111497

Fixes #109889

This PR adds `torch.export.export` as another `FXGraphExtractor` implementation. `torch.onnx.dynamo_export` automatically uses this new FX tracer when a `torch.export.ExportedProgram` is specified as `model`

Implementation is back compatible, thus non `ExportedProgram` models are handled the exact same way as before

### CODE REVIEW

File: torch/onnx/_internal/io_adapter.py [hunk 1]
```diff
@@ -487,3 +540,44 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
                 error_message="Model outputs incompatible with the format that was exported. ",
             )
         return flattened_outputs
+
+
+class PrependParamsAndBuffersAotAutogradInputStep:
+    """Prepend model parameters and buffers to the user input.
+
+    :func:`torch.export.export` lifts model parameters and buffers as model input, thus, they
+    must be added to the user input before the model  is executed.
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:549]:
```suggestion
    must be added to the user input before the model is executed.
```
nit

Comment (BowenBao) [reply torch/onnx/_internal/io_adapter.py:549]: Are the prepended params and buffers part of onnx model inputs? Do people need to supply them through input_dict for ort.run call?

File: torch/onnx/_internal/fx/torch_export_graph_extractor.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        updated_model_args = self.input_adapter.apply(
+            *model_args, model=model, **model_kwargs
+        )
+
+        # Export FX graph to ONNX ModelProto.
+        return self.pre_export_passes(options, model, model.graph_module, updated_model_args)  # type: ignore[return-value]
+
+    @_beartype.beartype
+    def pre_export_passes(
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/torch_export_graph_extractor.py:61]: Some passes still needed I believe, what is the plan for follow-ups?

File: torch/onnx/_internal/io_adapter.py [hunk 3]
```diff
@@ -487,3 +540,44 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
...
+
+    """
+
+    def apply(
+        self,
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+        model: Optional[
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:557]: pass model in `__init__` to keep `apply` signature unchanged?

File: torch/onnx/_internal/exporter.py [hunk 4]
```diff
@@ -350,16 +351,31 @@ class ResolvedExportOptions(ExportOptions):
...
     ):
-        if options is None:
-            options = ExportOptions()
+        from torch.onnx._internal.fx import (  # TODO: Prevent circular dep
+            diagnostics,
+            dynamo_graph_extractor,
+            torch_export_graph_extractor,
+        )
+
+        assert options is not None, "ExportOptions must be specified."
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:364]: why this change?

**Review by BowenBao — COMMENTED** [review]
A very quick skim through and some comments.

File: torch/onnx/_internal/io_adapter.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/io_adapter.py:549]: Yes, they are. They are supplied through `ort.run`

As an example, 

```python
        class Model(torch.nn.Module):
            def __init__(self):
                self.normal = torch.distributions.normal.Normal(0, 1)
                super().__init__()

            def forward(self, x):
                return self.normal.sample(x.shape)

        x = torch.randn(2, 3)
        exported_program = torch.export.export(Model(), args=(x,))
        _ = torch.onnx.dynamo_export(
            exported_program,
            x,
        )
```
results in a `GraphModule`

```python
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, arg0_1: f32[], arg1_1: f32[], arg2_1: f32[2, 3]):
            # File: /home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/distributions/normal.py:70, code: return torch.normal(self.loc.expand(shape), self.scale.expand(shape))
            expand: f32[2, 3] = torch.ops.aten.expand.default(arg0_1, [2, 3]);  arg0_1 = None
            expand_1: f32[2, 3] = torch.ops.aten.expand.default(arg1_1, [2, 3]);  arg1_1 = None
            normal: f32[2, 3] = torch.ops.aten.normal.Tensor_Tensor(expand, expand_1);  expand = expand_1 = None
            return (normal,)

Graph signature: ExportGraphSignature(parameters=[], buffers=['L__self___normal_loc', 'L__self___normal_scale'], user_inputs=['arg2_1'], user_outputs=['normal'], inputs_to_parameters={}, inputs_to_buffers={'arg0_1': 'L__self___normal_loc', 'arg1_1': 'L__self___normal_scale'}, buffers_to_mutate={}, backward_signature=None, assertion_dep_token=None)
Range constraints: {}
Equality constraints: []
```

with 1 user input and 2 buffers. When converting that as a ONN graph, we have all 3 as inputs:

```bash
[name: "arg0_1"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "arg1_1"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "arg2_1"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_value: 2
      }
      dim {
        dim_value: 3
      }
    }
  }
}
]
```

File: torch/onnx/_internal/fx/torch_export_graph_extractor.py (hunk 2 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/torch_export_graph_extractor.py:61]: Any in particular comes to mind?
I am adding them as I discover the need for them.  Right now there is only one simple test, so maybe some passes are missed

My idea was to create a PR in which we can reuse all `TestFxToOnnxWithOnnxRuntime` for testing with both `fx_tracer=dynamo_export_graph_extractor.DynamoExport` and `fx_tracer=torch_export_graph_extractor.TorchExport`. That would flush everything we need to support

File: torch/onnx/_internal/io_adapter.py (hunk 3 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/io_adapter.py:557]: I was wondering the same. I think this is a good idea.

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:364]: We were initializing `ExportOptions` in different points of the code. I found out this when writing a unit test that would detect an invalid `fx_tracer` and `model` combination. With this change, `def dynamo_export` is responsible to initialize `options` in a single spot

File: torch/onnx/_internal/exporter.py [hunk 5]
```diff
@@ -350,16 +351,31 @@ class ResolvedExportOptions(ExportOptions):
 
     @_beartype.beartype
     def __init__(
-        self, options: Optional[Union[ExportOptions, "ResolvedExportOptions"]]
+        self,
+        options: Optional[Union[ExportOptions, "ResolvedExportOptions"]] = None,
+        model: Optional[Union[torch.nn.Module, Callable, torch_export.ExportedProgram]] = None,  # type: ignore[name-defined]
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/exporter.py:356]: This I would like to pick your brain. Adding `model` solely for validation purposes. Should I move the validation outside `ResolvedExportOptions` and let the API simpler?

File: torch/onnx/_internal/exporter.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:364]: The original code looks clean too as all the None check are postponed to this point inside `ResolvedExportOptions`.

I don't have a strong opinion. If you are continuing with this change, please update argument `options: Optional[Union[ExportOptions, "ResolvedExportOptions"]] = None,` to `options: Union[ExportOptions, "ResolvedExportOptions"]`. Doesn't make sense to allow None that will immediately trigger assertion error.

File: torch/onnx/_internal/fx/torch_export_graph_extractor.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/torch_export_graph_extractor.py:59]: type promotion is immediate one I can think of. Yep we can play catch-up with test coverage and bench results.

**Review by BowenBao — APPROVED** [review]

File: torch/onnx/_internal/io_adapter.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/_internal/io_adapter.py]: I think we should put those back as initializers, can be in follow up.

Comment (thiagocrepaldi) [reply torch/onnx/_internal/io_adapter.py]: That is a good point. Will do it separately. 

Adding it back makes the ONNX file complete and executable by any runtime, but then we have the issue with models >2GB. On the other hand, if the user wants to save the onnx model (and discard `ExportOutput`), then all saved buffers/initializers are lost. Maybe we have to serialize `ExportOutput` in the  ExportedProgram` as input approach
